File: populating_right_pointers.py
# Time Complexity :
# O(N)

# Space Complexity :
# O(N)

# Did this code successfully run on Leetcode :
#Yes

#We do a BFS level order traversal of the tree. At each level, we make the connection between each of the nodes until the last node. We repeat this at all levels

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if root == None :
            return
        self.queue = collections.deque()
        count = 0

        self.queue.append(root)
        while len(self.queue) != 0 :
            size = len(self.queue)
            for i in range(0,size):
                curr = self.queue.popleft()
                if i != size-1 :
                    curr.next = self.queue[0]
                if curr.left != None :
                    self.queue.append(curr.left)
                    self.queue.append(curr.right)
        return root

File: populating-right-pointers.java

//sc is O(n)
//tc is O(n)
import java.util.*;

class Solution {
    public Node connect(Node root) {
        if (root == null)
            return null;

        Queue<Node> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int size = q.size();

            Node curr = q.poll();

            if (curr.left != null) {
                q.add(curr.left);
                q.add(curr.right);
            }

            for (int i = 0; i < size - 1; i++) {
                Node next = q.poll();
                curr.next = next;

                if (next.left != null) {
                    q.add(next.left);
                    q.add(next.right);
                }

                curr = next;
            }
        }

        return root;

    }
}
